LUCENE-10393: Unify binary dictionary and dictionary writer in kuromoji and nori

mocobeta opened the pull request:
Both Kuromoji and Nori have `BinaryDictionary` and `BinaryDictionaryWriter` classes, and there is significant code duplication. This PR unifies them by decoupling language-specific information (or morphological information) from the base dictionary interface.
For details, see https://issues.apache.org/jira/browse/LUCENE-10393

mocobeta committed:
factor out binary dictionary writer/reader for kuromoji

mocobeta committed:
apply to nori.

mocobeta committed:
lint

mocobeta committed:
add javadocs

rmuir commented:
I only looked at the high-level design so far, this seems to be a good approach @mocobeta ! Thank you for tackling it. I think the bottom-up approach is a good one, and splitting out the morphological data into separate interface makes sense to me. 

I would suggest reconsidering the name `MorphAttributes`, mostly because "Attributes" already has a complex meaning within lucene analysis. Some possibilities (not exhaustive list):
* `MorphData`
* `DictionaryData`

I will do more review and testing, I am digging into it in detail.

rmuir commented:
I ran `./gradlew regenerate --rerun-tasks` on your branch as an additional test and all binary data files were unchanged. So I feel good about correctness!

mocobeta commented:
> I ran ./gradlew regenerate --rerun-tasks on your branch as an additional test and all binary data files were unchanged. So I feel good about correctness!

Thanks, @rmuir for confirming that! This is code refactoring and shouldn't change the outputs of the BinaryDictionaryWriter.

About the interface,  `MorphData` sounds fine to me if `Attributes` is confusing. I'll rename the classes.

mocobeta committed:
rename MorphAttributes to MorphData

mocobeta committed:
minor refactor on dictionary writer

mocobeta committed:
lint

mocobeta committed:
ConnectionCosts and ConnectionCostsWriter can be shared in a similar way.

mocobeta commented:
We could have a common `DictionaryBuilder` class in analyzers-common but it brings too complex class hierarchy to me. I'd postpone refactoring XXXDictionaryBuilder until we come up with good interfaces or framework for that - it may need public interface changes and is out of the scope of this PR.

rmuir commented on lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/util/UnknownDictionaryWriter.java:
```diff
@@ -56,9 +59,14 @@ public void putInvokeDefinition(String characterClassName, int invoke, int group
   }
 
-  @Override
+  // @Override
```
was this intentional?

mocobeta replied:
No, it looks the interfaces were still work-in-progress.
I think an abstract method should be added to the base dictionary writer class. https://github.com/apache/lucene/pull/740/commits/7d4351417a0bb96e647525cae6eb8124e0805016

mocobeta committed:
add write(Path) abstract method to BinaryDictionaryWriter

mocobeta committed:
CharacterDefinition/CharacterDefinitionWriter can be shared in a similar way.

mocobeta committed:
small refactoring on codec header constants

mocobeta committed:
remove obsolete comments

rmuir commented on lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/UnknownMorphData.java:
```diff
@@ -0,0 +1,44 @@
+import org.apache.lucene.util.IOSupplier;
+
+/** Morphological information for unk dictionary. */
+public class UnknownMorphData extends TokenInfoMorphData {
```
We can remove the `public` keyword here I think? (Same for the nori equivalent)

mocobeta replied:
Thanks, yes - they also can be final. https://github.com/apache/lucene/pull/740/commits/b1a0033ef118bef0300ad61db036a2d3dd609fa1

mocobeta committed:
reduce visibility of XXXMorphData; and they can be final

mocobeta committed:
add javadocs

mocobeta commented:
I think I finished what I'd like to make changes in kuromoji and nori.
Now, `ConnectionCosts`/`ConnectionCostsWriter` and `CharacterDefinition`/`CharacterDefinitionWriter` are also reside in `analyzers-common` in a very similar way to `BinaryDictionary`/`BinaryDictionaryWriter`.

I manually checked the shared code is exactly the same for kuromoji and nori in the current `main` (in other words, they were copied and have never been changed so far). Also, `gradlew regenerate --rerun-tasks` does not change the generated dictionary files for me.

rmuir commented on lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/util/TokenInfoDictionaryEntryWriter.java:
```diff
@@ -0,0 +1,221 @@
+import org.apache.lucene.util.ArrayUtil;
+
+/** Writes system dictionary entries */
+public class TokenInfoDictionaryEntryWriter extends DictionaryEntryWriter {
```
This also looks like it need not be public? I'm just looking for opportunities to reduce visibility as it really helps simplify the javadocs output on users.

mocobeta replied:
Ah, yes!

mocobeta committed:
reduce visibility of DictionaryEntryWriters

mocobeta committed:
add documentation about ipadic format

mocobeta committed:
putEntry() also can be protected

mocobeta commented:
The core dictionary logic is split into two modules (analysis-common and analysis-kuromoji/nori), I manually tested the tokenizers work with Java modules - for now luke app is a quick way to do so. It might be good to have some automated tests. I will try to locate where we should have such tests, maybe in another patch.
I think these changes are module-friendly, but just in case.

uschindler commented:
I like the idea to remove the code duplication and have only one implementation.

On the other hand, if you look at LOC before/after: +1,818 −1,492
We now have 326 lines more, in addition both NORI and Kuromojo hardly depend on code in a different module (analysis-common), which now gets public.

So I feel a bit ambiguous about if it really makes sense to forcefully combine those implementations just because we have an overlap of 200 lines.

mocobeta commented:
About the increased number of lines, the majority of them are license headers and documentation.
There are added 22 files (to sort out interfaces) and removed 4 files - each license header contains 15 lines so 288 license headers are added. Another reason for increasing the number of lines is newly introduced interfaces and their Javadocs. I think the substantial amount of code of the implementation classes was reduced by this change, though, I didn't count it.

I think the apparent demerit of this patch is exposing dictionary internals as public interfaces (and kuromoji and nori depend on it). We would have to choose which is better - keep hiding internals and maintain duplicated code, or open up some internals and share them. I myself would prefer the latter approach to ease the development or bug fixes that are common to kuromoji and nori, and prevent further diversifying them.

uschindler commented:
> I think the apparent demerit of this patch is exposing dictionary internals as public interfaces (and kuromoji and nori depend on it).

This is also my biggest concern, although the additional complexity and additional interfaces also make me sad.

One possibility is to use module system to at least hide the interface for "modern users" (are there any using module system): hide the package :-)

mocobeta commented:
I just wanted to explain my view about interfaces. From my perspective, I don't think this adds complexities to interfaces.
As I wrote in the Jira issue, there are only two conceptual interfaces.

- Dictionary: a high-level interface parameterized by a specific MorphAttributes
- MorphAttributes: a high-level interface that represents morphological information. This is supposed to be extended to hold language-specific details.

Our current kuromoji/nori interfaces mix up "dictionary-lookup" and "language-specific feature", and in theory - they should be decoupled as original [mecab](https://github.com/taku910/mecab) library and its various ports do so in order to have only one analysis engine that can "switch" specific dictionaries.

uschindler commented:
> Our current kuromoji/nori interfaces mix up "dictionary-lookup" and "language-specific feature", and in theory - they should be decoupled as original [mecab](https://github.com/taku910/mecab) library and its various ports do so in order to have only one analysis engine that can "switch" specific dictionaries.

The module structure could also be refactored: Why need analyzer modules per language? Just have one morphological module with pluggable implementations.

mocobeta commented:
> Just have one morphological module with pluggable implementations.

**In theory**, we should have one analysis engine and kuromoji and nori should be the language-specific "dictionaries". I believe LUCENE-8816 will open up the possibility, but we have a long way to go...

uschindler commented:
So lets merge this PR. At a later stage we should collect all those refactored classes and put them in a separate module "analysis-morph".

uschindler commented:
Maybe export `org.apache.lucene.analysis.morph` only to kuromoji and nori. The classes inside there are *all* private? Only the actual implementations should be visible?

mocobeta commented:
> Maybe export org.apache.lucene.analysis.morph only to kuromoji and nori. 

First I wanted and tried to do so, but in order to export "org.apache.lucene.analysis.morph" package only to them, analysis-common requires kuromoji and nori otherwise it does not know them at all? Maybe I'm missing something?

mocobeta commented:
> For now, I saw compilation fails with warning when I add to clause to exports.

I temporarily suppressed this warning by `@SuppressWarnings("module")`, then the whole project passed compilation with this `to` clause.
```
exports org.apache.lucene.analysis.morph to org.apache.lucene.analysis.kuromoji, org.apache.lucene.analysis.nori;
```

While the built kuromoji and nori jars works fine, I found an interesting issue. It seems we also have to modularize tests of kuromoji / nori (currently, the tests are "unnamed" module so `org.apache.lucene.analysis.morph` package is not visible to them).
```
./gradlew -p lucene/analysis/kuromoji/ compileJava --rerun-tasks
BUILD SUCCESSFUL in 10s
```
```
./gradlew -p lucene/analysis/kuromoji/ test

> Task :lucene:analysis:kuromoji:compileTestJava FAILED
/mnt/hdd/repo/lucene/lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/TestExternalDictionary.java:20: error: package org.apache.lucene.analysis.morph is not visible
import static org.apache.lucene.analysis.morph.BinaryDictionary.DICT_FILENAME_SUFFIX;
                                        ^
  (package org.apache.lucene.analysis.morph is declared in module org.apache.lucene.analysis.common, which does not export it to the unnamed module)

```

rmuir commented:
along the same lines of visibility, I think we actually hurt ourselves the way we package things in kuromoji and nori. The unnecessary subpackages force a lot of internal stuff to be public that shouldn't be.

For example: `TokenInfoMorphData` shouldn't be publicly visible. But it has to be because it is in the `.dict` package and something from `.util` uses some constants from it.

we should be packaging our code, in a way that is simplest to users (least number of public classes).

mocobeta commented:
> For example: TokenInfoMorphData shouldn't be publicly visible. But it has to be because it is in the .dict package and something from .util uses some constants from it.

The current way of packaging around dictionaries, `.dict` and `.util`, in kuromoji and nori seem awkward to me too. We could merge them or refine the package structure on the main branch, maybe in another issue.

mocobeta commented:
I looked at the packages again. It'd make sense to me to simply move all classes in `.util` package (dictionary builder and writers) to `.dict` so that classes in `.dict` can be kept package-private, and remove `.util` package. Can we make the package restructure in this PR? Moving classes to another package breaks backward compatibility, but I suppose we are not going to backport this to 9x anyway.

rmuir commented:
we could do that stuff in another PR. there is enough changes in this PR already I think? And the problem is really a separate, existing problem...

mocobeta committed:
merge .util package to .dict

mocobeta committed:
Revert "merge .util package to .dict"

This reverts commit 67ed0169733d3fcde47946f89dd6abac5fb31bce.

mocobeta commented:
I accidentally committed https://github.com/apache/lucene/pull/740/commits/67ed0169733d3fcde47946f89dd6abac5fb31bce and reverted it. It merges `.util` to `.dict` package, minimize the visibility of the internal classes to package-private, and deletes `.util` package.
It'd be easy refactoring (with the help of IDE) but many files are affected. I agree that it should be done in another issue/PR.

mocobeta committed:
add module tests for kuromoji and nori

mocobeta committed:
add changes entry.

mocobeta committed:
Merge branch 'main' into jira/lucene-10393-refine-dictionary-api

mocobeta commented:
To me, this is already self-contained and ready to be merged. This is not perfect though, I think it would be a start point to move forward (having flexibly switchable or modularized dictionaries, or unified Tokenizers at some level so that we can simultaneously improve/optimize both of kuromoji and nori; I'll keep going to work on it once this is successfully merged).

I added the CHANGES entry but I'd need approvals to merge such a large patch. I understand this perhaps could be a bit controversial, I will keep it open to wait for feedback from others.

mocobeta commented:
Hi @rmuir, I just wanted to ping you to let you know that I requested a review (in github) a little while ago. I don't intend to rush you, thanks.

rmuir reviewed (approved):
thanks for this refactoring. it is a good first step!

rmuir commented:
also, sorry about the review slowness. i didn't want to just click "approve" without taking another pass through the comments and code. Again, I like the way the concerns were split apart, the explanation you gave about +/- LOC from github is exactly how I feel, too.

The overall algorithm is the same one here for nori and kuromoji, so it is a shame that we have duplicated implementation code (the holy grail will be factoring the actual tokenization logic!). At the same time, different languages have quirks about them and need different encoding/compression to be efficient. Different dictionaries might have quirks, too. It would be great to give all reasonable options compatible with the apache2 license to the user, without forking thousands of lines of Tokenizer code, each time.

uschindler reviewed (approved):
I agree with Robert's comment.

uschindler commented:
> I added test modules `analysis/kuromoji.tests` and `analysis/nori.tests` to make sure that both tokenizers correctly load the dictionary resources and work in module-mode. They are tiny tests but it'd be good to have ones for sanity checks.

This is not needed due to the improved TestRandomChains. This one instantiates all analyzer compients from module system.

uschindler commented:
As said before, I would not add the extra module tests. We have the integration test for all analyzers, so if tokenizer can be instantiated it is happy. If there is an IOException or similar it would fail.

I don't like the many Gradle modules, but I leave it up to you if you want to keep the tests.

mocobeta committed:
Revert "add module tests for kuromoji and nori"

This reverts commit c069ecb69b7390b84099cca3be65dad6cb879973.

mocobeta commented:
>>  I added test modules analysis/kuromoji.tests and analysis/nori.tests to make sure that both tokenizers correctly load the dictionary resources and work in module-mode. They are tiny tests but it'd be good to have ones for sanity checks.

> This is not needed due to the improved TestRandomChains. This one instantiates all analyzer compients from module system.

I removed the test modules. If we need specific tests for kuromoji or nori at some point in the future, then we will be able to re-add them.

mocobeta commented:
@rmuir @uschindler 
Thank you for your thorough review! I thought it'd take some more time.

I'm merging it now - I think it'd be better to open a follow-up issue to make `org.apache.lucene.analysis.morph` package visible only to kuromoji and nori.

uschindler commented:
FYI, here is this Test (it also checks if all components have factories): https://github.com/apache/lucene/tree/main/lucene/analysis.tests/src/test/org/apache/lucene/analysis/tests

Of course RandomChains won't trigger on every run, because it randomly builds combinations of tokenizers and filters, but a broken Tokenizer Component that can't load its resources will for sure break this test.

mocobeta commented:
Just a quick note: I opened #772, which slightly improves the encapsulation of the dictionary-related internals. The whole refactoring was done by IDE and I think there wouldn't be big matters to be discussed, I am going to merge it tomorrow.

uschindler commented:
Yes this was intentional. It breaks API.
